### DISCUSSION TITLE
json binary and collation

### DISCUSSION THREAD

**Discussion by carlituxman** [spatie/laravel-translatable#349] [Q&A] [answered]
When use json field on mysql hasn't collation because is binary, but in mariadb json migrate create a longText field with collation.

We have a utf8mb4_unicode_ci (case insensitive)

Then doing a where like %$search% with $search = 'sesión', $search = 'sesion', $search = 'Sesión' should return same results, with mariadb works, but mysql not. We needed extends Illuminate\Database\Eloquent\Builder with this:

`public function whereCollateJsonLike($column, $value = null, $or = false)
    {
        $operator = 'like';
        $searchOperator = $or ? 'orWhereRaw' : 'whereRaw';

        $collation = $this->getConnection()->getConfig()['collation'];
        $appLocale = app()->getLocale();

        $searchPattern = '%' . $value . '%';
        
        // return self::{$searchOperator}($column . '->"$.' . $appLocale . '" ' . $operator . '?', $searchPattern);
        return self::{$searchOperator}($column . '->"$.' . $appLocale . '" COLLATE ' . $collation . ' ' . $operator . '?', $searchPattern);
    }`

To specify collation on search. too much tangled

I share with you because it's been a headache

**Comment by sotospez** [comment, marked as answer]
Hey @carlituxman, you are spot on about the MySQL binary JSON behavior vs MariaDB!

I ran into the exact same issue while dealing with Greek characters and accents (e.g., searching for uppercase/lowercase with accents like Σωτήρης vs σωτηρης vs ΣΩΤΗΡΗΣ).

To build on top of your solution here is a Trait version. It solves this exact headache for both MySQL and PostgreSQL handles dynamic locales (so you can loop through multiple languages in a single search) and uses JSON_UNQUOTE to clean up the raw string quotes:

```php
use Illuminate\Database\Eloquent\Builder;

trait HasTranslatableSearch
{
    /**
     * Scope to perform a case and accent insensitive search on translatable (JSON) fields.
     * Supports MySQL and PostgreSQL with proper collations/operators.
     *
     * @param Builder $query
     * @param string $column The column name (e.g., 'title')
     * @param string $locale The locale key (e.g., 'el' or 'en')
     * @param string $value The search term
     * @return Builder
     */
    public function scopeOrWhereTranslatable(Builder $query, string $column, string $locale, string $value): Builder
    {
        $driver = $query->getConnection()->getDriverName();

        if ($driver === 'mysql') {
            $collation = config('database.connections.mysql.collation', 'utf8mb4_unicode_ci');

            return $query->orWhereRaw(
                "JSON_UNQUOTE(JSON_EXTRACT(`{$column}`, '$.{$locale}')) COLLATE {$collation} LIKE ?",
                ["%{$value}%"]
            );
        }

        if ($driver === 'pgsql') {

            return $query->orWhereRaw(
                "\"{$column}\"->>'{$locale}' ILIKE ?",
                ["%{$value}%"]
            );
        }

        // Fallback for SQLite or other drivers using standard custom extraction
        return $query->orWhereRaw(
            "JSON_UNQUOTE(JSON_EXTRACT(`{$column}`, '$.{$locale}')) LIKE ?",
            ["%{$value}%"]
        );
    }
}
```
Usage in Controller (looping through multiple locales):
```php
$query->where(function ($qb) use ($search) {
    foreach (['en', 'es', 'el'] as $locale) {
        $qb->orWhereTranslatable('title', $locale, $search);
    }
});
```
Thanks for sharing your initial code.
